File: platform/polyglot/rag_search_algorithms/py/rag_search_algorithms/rrf.py
"""Reciprocal Rank Fusion (RRF) algorithm."""

from typing import List, Tuple


def reciprocal_rank_fusion(
    vector_results: List[Tuple[str, float]],
    bm25_results: List[Tuple[str, float]],
    alpha: float,
    k_rrf: int = 60,
) -> List[Tuple[str, float]]:
    """Weighted Reciprocal Rank Fusion.

    Combines vector and BM25 search results using RRF scoring.

    Args:
        vector_results: List of (content_hash, score) from vector search.
        bm25_results: List of (content_hash, score) from BM25 search.
        alpha: Weight for vector results (1-alpha for BM25). Range [0, 1].
        k_rrf: RRF constant (default 60). Higher values flatten rank differences.

    Returns:
        Sorted list of (content_hash, fused_score) in descending score order.
    """
    scores: dict[str, float] = {}
    for rank, (ch, _score) in enumerate(vector_results):
        scores[ch] = scores.get(ch, 0.0) + alpha * (1.0 / (k_rrf + rank + 1))
    for rank, (ch, _score) in enumerate(bm25_results):
        scores[ch] = scores.get(ch, 0.0) + (1.0 - alpha) * (1.0 / (k_rrf + rank + 1))
    return sorted(scores.items(), key=lambda x: x[1], reverse=True)

File: platform/polyglot/rag_search_algorithms/mjs/src/rrf.mjs
/**
 * @fileoverview Reciprocal Rank Fusion (RRF) algorithm.
 */

/**
 * Weighted Reciprocal Rank Fusion.
 *
 * Combines vector and BM25 search results using RRF scoring.
 *
 * @param {Array<[string, number]>} vectorResults - (contentHash, score) from vector search
 * @param {Array<[string, number]>} bm25Results - (contentHash, score) from BM25 search
 * @param {number} alpha - Weight for vector results (1-alpha for BM25). Range [0, 1].
 * @param {number} [kRrf=60] - RRF constant. Higher values flatten rank differences.
 * @returns {Array<[string, number]>} Sorted (contentHash, fusedScore) in descending order.
 */
export function reciprocalRankFusion(vectorResults, bm25Results, alpha, kRrf = 60) {
  /** @type {Map<string, number>} */
  const scores = new Map();

  for (let rank = 0; rank < vectorResults.length; rank++) {
    const [ch] = vectorResults[rank];
    const prev = scores.get(ch) || 0;
    scores.set(ch, prev + alpha * (1.0 / (kRrf + rank + 1)));
  }

  for (let rank = 0; rank < bm25Results.length; rank++) {
    const [ch] = bm25Results[rank];
    const prev = scores.get(ch) || 0;
    scores.set(ch, prev + (1.0 - alpha) * (1.0 / (kRrf + rank + 1)));
  }

  return [...scores.entries()].sort((a, b) => b[1] - a[1]);
}
